Feature/rocs scoring

fulopjoz opened the pull request:
# Add ROCS-based molecular similarity scoring

## Summary
Implements ROCS (Rapid Overlay of Chemical Structures) scoring for shape-based molecular similarity evaluation in DrugEx reinforcement learning workflows.

## Changes
- Added 3 new ROCS scorers with multiple backend support (OpenEye, RDKit, CDPKit)
- Added 4 conformer generators (Omega, ETKDGv3, ConfGen, confgenx)
- Added initial tutorial for pipeline validation
- Implemented multi-query support with  best-score selection
- Added parallel processing

## Usage
```python
from drugex.training.scorers.cli_rocs import CLIROCSScorer
from drugex.training.scorers.conformer_generators import OmegaConformerGenerator

conformer_gen = OmegaConformerGenerator()
scorer = CLIROCSScorer(
    conformer_generator=conformer_gen,
    query_files={"query1": "model1.sq"},
    score_type="TanimotoCombo"
)
```

## Dependencies
- OpenEye toolkits, CDPKit, RDKit


@martin-sicho @HellevdM

fulopjoz committed:
add rocs documentation path to .gitignore

fulopjoz committed:
add FastROCS-based scorer implementation

fulopjoz committed:
implement molecule filtering and scoring enhancements in FastROCS scorer

fulopjoz committed:
enhance FastROCS scorer with optimized memory management and adaptive resource utilization

fulopjoz committed:
add FastROCS scorer implementation with optimized memory management and adaptive resource utilization

fulopjoz committed:
add new FastROCS scorer implementation with conformer generation and scoring capabilities

fulopjoz committed:
Implement new feature for user authentication and improve error handling

fulopjoz committed:
add new binary file for model3-4 version 1 in FastROCS tutorial

fulopjoz committed:
Enhance ROCS function to support GPU acceleration and improve input handling for RDKit molecules

fulopjoz committed:
Add detailed docstring for OpenEyeScorer initialization parameters and update licensing information

fulopjoz committed:
Refactor OMEGA and ROCS functions for improved file handling and error management; increase conformer sampling and optimize scoring process.

fulopjoz committed:
Improve worker calculation and error handling in FastROCS; adjust CPU core allocation, enhance memory pool initialization, and standardize logging intervals.

fulopjoz committed:
Enhance OMEGA and ROCS functions with configurable parameters for conformer generation and threading; improve error handling and memory management.

fulopjoz committed:
notebook with examples or rocs scorers and drugex implementation  pipeline

fulopjoz committed:
Enhance memory management in ez_rocs module; initialize OpenEye memory pool at import and improve error handling during initialization.

fulopjoz committed:
Enhance FastROCS module with improved memory management and optimized filtering criteria; initialize OpenEye memory pool at import, adjust worker calculations, and streamline database preparation for better performance.

fulopjoz committed:
Reduce verbosity in GPU preparation logging; only log significant discrepancies in processed molecule counts.

fulopjoz committed:
completed pipeline on fastrocs scorer

fulopjoz committed:
Add GPU performance comparison script for ROCS scorers

fulopjoz committed:
Add ROCS scorers comparison script to benchmark ez_rocs and fastrocs implementations

fulopjoz committed:
Add benchmark scripts for ez_rocs and fastrocs GPU performance comparison

fulopjoz committed:
Add detailed README for DrugEx-ROCS scorers with configuration and usage examples

fulopjoz committed:
folder with ipynb output from example drugex-rocs run

fulopjoz committed:
some ideas how to make the pipeline RL faster ??

fulopjoz committed:
remove old version ipynb

fulopjoz committed:
(fix) rdkit-pypi changed to rdkit due compatibility

fulopjoz committed:
(fix) added sources from documenation and minor changes

fulopjoz committed:
(feat) paralell processing of the OMEGA conformer generation !! not tested !!

fulopjoz committed:
remove file

fulopjoz committed:
Remove fastrocs_v2_scorer.py file as part of codebase cleanup.

fulopjoz committed:
Enhance DrugEx_RNN_ROCS_v2 notebook with improved GPU processing details and scoring metrics. Updated scorer parameters for better performance and added parallel processing capabilities - which is in testing and not reliable and fully functional yet!!!!!!!. Adjusted output formatting

fulopjoz committed:
new run data

fulopjoz committed:
Increase stdout maxBuffer to prevent state retrieval failures

Raise the maxBuffer limit for stdout to avoid "maxBuffer length exceeded" errors when retrieving the working state. This ensures reliable state fetching even with large outputs.

fulopjoz committed:
Implement adaptive model selection in OpenEyeScorer for improved performance tracking and scoring. Enhance handling of multiple model paths and validate query files. Introduce parallel execution options for scoring, optimizing resource utilization.

fulopjoz committed:
Implement AdaptiveModelSelector for dynamic model performance tracking in ROCS scoring. Enhance query file handling and validation, allowing for both single and multiple model inputs. Improve scoring methods with parallel and sequential execution options, ensuring maximum scores are computed efficiently across models.

fulopjoz committed:
Enhance FastROCS scoring with GPU memory management, adaptive batch sizing, and shared molecule caching. Introduce functions for monitoring memory usage and optimizing processing based on available resources. Implement pre-filtering of molecules to improve scoring efficiency and reduce memory pressure during batch processing.

fulopjoz committed:
Remove fastrocs_paralell.py file as part of codebase cleanup.

fulopjoz committed:
add (multiple models .sq files scoring provided in list)
refactor FastROCS scoring module by removing unused functions and classes related to GPU memory management, molecule complexity calculation, and batch processing. Streamline the scoring process for improved efficiency and maintainability.

fulopjoz committed:
Add new ROCS model .sq files

fulopjoz committed:
tested list of .sq files

fulopjoz committed:
Enhance AdaptiveModelSelector documentation with detailed explanations of Exponential Moving Average (EMA) usage and its impact on model performance tracking. Add TODOs for future improvements including tuning parameters and implementing a warm-up period.

fulopjoz committed:
Update finetune_fit, finetune_smiles, ligand_test, ligand_train, and rocs_agent_smiles files with new training data and results. Enhance data consistency and expand the dataset for improved model performance tracking.

fulopjoz committed:
Update rocs_agent_fit.tsv with new training results, replacing previous data to reflect improved model performance metrics. Adjusted values for various epochs to enhance accuracy in tracking training progress.

fulopjoz committed:
Update AdaptiveModelSelector documentation with a new reference link to relevant research, enhancing context for model performance tracking.

fulopjoz committed:
Update .gitignore to include new entries for drugex_rocs_backup_files and tutorial/rocs/docs, ensuring proper exclusion of backup and documentation files from version control.

fulopjoz committed:
Add new ROCS model files and trypsin data

-  three new ROCS model files (model1, model2, model3)  results from vrocs.
-  binary files for trypsin data, sq files form vROCS.

fulopjoz committed:
rl with cli model

fulopjoz committed:
Refactor ez_rocs.py and fastrocs.py for improved performance and memory management

- Updated AdaptiveModelSelector for better model selection and performance tracking.
- Enhanced OMEGA and ROCS functions for improved conformer generation and scoring.
- Removed outdated readme_rocs.md to streamline documentation.
- Implemented better error handling and resource management in FastROCS scoring.
- Adjusted filtering thresholds and batch sizes for more efficient processing.

fulopjoz committed:
Add new ROCS API scorer and batch analysis scripts

- Introduced `api_base_rocs.py` for shape-based scoring using OpenEye FastROCS.
- Added `batch_rocs_analysis.py` to compare CLI and API scoring implementations.
- Created `cli_base_rocs.py` for CLI-based ROCS scoring.
- Implemented `rocs_results_analyzer.py` for analyzing and validating ROCS results.
- Included comprehensive scoring results for trypsin data across multiple implementations.
- Added detailed analysis summaries and visualizations for performance comparison.

fulopjoz committed:
rm dockstring

fulopjoz committed:
Add Smart CLI ROCS Scorer with Multi-Query Support

- Introduced `SmartCLIROCSScorer` for scoring using CLI ROCS.
- Supports multiple .sq query files and selects the best score.
- Implements batch processing for large datasets and integrates with DrugEx.
- Includes comprehensive error handling and molecule preparation methods.
- Validates query files and provides detailed scoring results.

fulopjoz committed:
Update DrugEx RNN ROCS tutorial notebook with CLI ROCS scorer and implementation into DrugEx

fulopjoz committed:
feat: integrate SmartCLI scorer with memory monitoring

- Add SmartCLIROCSScorer with caching enabled
- Fix AttributeError with FileMonitor setup
- Add memory usage tracking throughout training
- Display cache statistics and cleanup resources

fulopjoz committed:
feat: add ROCS scorers comparison test suite

fulopjoz committed:
feat: add SmartCLIROCSScorer with memory optimization

- Thread-safe LRU conformer cache with corruption detection
- Process isolation for large batches (>500 molecules)
- Memory pressure monitoring and automatic cleanup
- Multi-query support with best score selection
- Conservative settings for RL training stability

fulopjoz committed:
feat: enhance OMEGA and ROCS functions for improved conformer handling

- Updated OMEGA function to include save directory parameter for output files.
- Modified ROCS function to accept a save directory for results.
- Enhanced RocsScorer class to manage conformer saving and temporary directory creation.
- Improved cleanup process for temporary files and directories.
- Added detailed docstrings for better parameter descriptions and return values.

fulopjoz committed:
feat: update DrugEx RNN ROCS tutorial for memory optimization

fulopjoz committed:
fix: update memory usage statistics in ROCS tutorial notebook

fulopjoz committed:
feat: update ROCS RL data and training metrics
.

HellevdM committed:
print extra errors, pass through all config settings to _isolated_scoring_workers, add suffix to scorer key

HellevdM committed:
simplify and speed up cli rocs scorer

HellevdM committed:
bring back gpu option

HellevdM committed:
remove unkown function

fulopjoz committed:
Merge pull request #3 from fulopjoz/fix/cli

HellevdM committed:
add optional multi-task scorer rocs

HellevdM committed:
Merge remote-tracking branch 'origin/master' into feature/multitask_rocs

HellevdM committed:
handle no valid molecules

HellevdM committed:
log epochs in sequence_explorer policy gradient

HellevdM committed:
add optional loss_tolerance to fit for finetuning

HellevdM committed:
add option of using sdf files as queries

HellevdM committed:
remove memory manager

HellevdM committed:
separate conformer generation from cli rocs scorer

HellevdM committed:
move conformer generation to separate file and add Schrodinger conformer generator

HellevdM committed:
update documentation CLIROCSScorer

HellevdM committed:
small fix in confomer generation import and rocs command build

fulopjoz committed:
chore: update gitignore for ROCS development files

- Ignore training result directories (rl_runs, finetune_runs, encoded_rnn)
- Ignore backup files and test scripts
- Ignore diagnostic utilities

fulopjoz committed:
feat(scorers): add RDKit ETKDGv3  and CDPKit Congen conformer generators

fulopjoz committed:
feat(scorers): add open-source ROCS implementations

Add CDPKit ROCS scorer:
- CDPKitROCSAggregateScorer for multi-reference scoring
- CDPKitROCSSupermoleculeScorer for single supermolecule
- GaussianShape alignment with TanimotoCombo scoring
- Match CDPKit shapescreen CLI defaults for consistency
- Support parallel processing with multiprocessing

Add RDKit ROCS scorer:
- RDKitAggregateScorer using rdShapeAlign
- RDKitSupermoleculeScorer for supermolecule queries
- Shape and color Tanimoto similarity scoring
- Support parallel scoring with worker pools

fulopjoz committed:
fix(scorers):  Fix multi-query file handling ( len(self.queries) -> len(query_files))

fulopjoz committed:
remove deprecated ROCS implementations

fulopjoz committed:
docs(tutorial): add ROCS pipeline validation notebook

- add initial ROCS integration pipeline check with DrugEx RNN
- include reference data and training comparison plots
- validate ROCS scorer implementations (OpenEye, CDPKit, RDKit)
- foundation for workflow documentation

fulopjoz committed:
chore: remove unused ROCS model files

fulopjoz committed:
refactor(scorers): unify RDKit ROCS scorer implementations

  Consolidate RDKitAggregateScorer and RDKitSupermoleculeScorer into single  RDKitROCSScorer class. Legacy classes preserved with deprecation warnings.

fulopjoz committed:
refactor(scorers): unify CDPKit ROCS scorer implementations

  Consolidate CDPKitROCSAggregateScorer and CDPKitROCSSupermoleculeScorer into   single CDPKitROCSScorer class. Legacy classes preserved with deprecation warnings.

  -Next: remove deprecated classes

fulopjoz committed:
Merge branch 'dev' into feature/rocs-scoring

fulopjoz committed:
chore: clean up .gitignore by removing obsolete entries for ROCS test scripts and backup files

fulopjoz committed:
Add comprehensive documentation, interactive viewer, and diagnostic tools
for comparing ROCS pipeline implementations and optimizing molecular
similarity workflows.

fulopjoz committed:
chore: remove obsolete ROCS tutorial scripts

fulopjoz committed:
refactor(scorers): simplify CDPKit ROCS implementation and remove deprecated wrappers

BREAKING CHANGES:
- Remove CDPKitROCSAggregateScorer and CDPKitROCSSupermoleculeScorer (use CDPKitROCSScorer)
- Change getKey() return type from str to List[str] for consistency with scorer interface

Changes:
- Consolidate verbose documentation into concise docstrings
- Inline _score_cdpkit_conformers() logic into getScores() method
- Remove explanatory comments while preserving all scoring functionality
- Maintain identical TanimotoCombo scoring algorithm and optimization parameters

Migration:
# Old code
scorer = CDPKitROCSAggregateScorer(gen, ["ref.sdf"])
# New code
scorer = CDPKitROCSScorer(gen, ["ref.sdf"])

fulopjoz committed:
refactor(scorers): simplify RDKit ROCS implementation and remove deprecated wrappers

BREAKING CHANGES:
- Remove RDKitAggregateScorer and RDKitSupermoleculeScorer (use RDKitROCSScorer)
- Change getKey() return type from str to List[str] for consistency with scorer interface

Changes:
- Consolidate verbose documentation into concise docstrings
- Extract SMILES conversion into dedicated _convert_to_smiles() method
- Rename _detect_supermol_mode() to _is_single_reference() for clarity
- Streamline shape alignment logic and exception handling
- Maintain identical TanimotoCombo scoring algorithm

Migration:
# Old code
scorer = RDKitAggregateScorer(gen, [mol1, mol2])
# New code
scorer = RDKitROCSScorer(gen, [mol1, mol2])

fulopjoz committed:
feat(scorers): implement stereoisomer enumeration in CDPKit conformer generator

BREAKING CHANGE: Now generates conformers for multiple stereoisomers,
matching RDKit/OpenEye behavior. Output increased from max_conformers
to (max_centers × max_conformers) structures per molecule.

- Add StereoisomerGenerator for explicit stereoisomer enumeration
- Rewrite _get_isomers() to enumerate up to max_centers stereoisomers
- Update documentation to reflect new behavior
- Align CDPKit conformer generation with other implementations

fulopjoz committed:
docs(tutorial): add comprehensive ROCS-RNN demo notebook for CCR2 ligand design

Add complete end-to-end tutorial demonstrating DrugEx with shape-based ROCS scoring.
Compares RDKit, CDPKit, and OpenEye ROCS implementations in multi-objective RL workflow.

Features:
- Fine-tuning pretrained RNN on CCR2 ligand data
- Multi-objective RL combining ROCS shape similarity and synthetic accessibility
- Molecule generation with all three ROCS backends

fulopjoz committed:
refactor(scorers): simplify RDKit ROCS implementation and remove deprecated wrappers

- Remove RDKitAggregateScorer and RDKitSupermoleculeScorer (use RDKitROCSScorer)
- Change getKey() return type from str to List[str] for consistency with scorer interface

- Consolidate verbose documentation into concise docstrings
- Extract SMILES conversion into dedicated _convert_to_smiles() method
- Rename _detect_supermol_mode() to _is_single_reference() for clarity
- Streamline shape alignment logic and exception handling
- Maintain identical TanimotoCombo scoring algorithm

Migration:
# Old code
scorer = RDKitAggregateScorer(gen, [mol1, mol2])
# New code
scorer = RDKitROCSScorer(gen, [mol1, mol2])

fulopjoz committed:
feat(scorers): implement stereoisomer enumeration in CDPKit conformer generator

Now generates conformers for multiple stereoisomers, matching RDKit/OpenEye behavior. Output increased from max_conformers  to (max_centers × max_conformers) structures per molecule.

- Add StereoisomerGenerator for explicit stereoisomer enumeration
- Rewrite _get_isomers() to enumerate up to max_centers stereoisomers
- Update documentation to reflect new behavior
- Align CDPKit conformer generation with other implementations

fulopjoz committed:
refactor(scorers): simplify CDPKit ROCS implementation and remove deprecated wrappers

- Remove CDPKitROCSAggregateScorer and CDPKitROCSSupermoleculeScorer (use CDPKitROCSScorer)
- Change getKey() return type from str to List[str] for consistency with scorer interface

- Consolidate verbose documentation into concise docstrings
- Inline _score_cdpkit_conformers() logic into getScores() method
- Remove explanatory comments while preserving all scoring functionality
- Maintain identical TanimotoCombo scoring algorithm and optimization parameters

Migration:
# Old code
scorer = CDPKitROCSAggregateScorer(gen, ["ref.sdf"])
# New code
scorer = CDPKitROCSScorer(gen, ["ref.sdf"])

fulopjoz committed:
docs(tutorial): add comprehensive ROCS-RNN demo notebook for CCR2 ligand design and remove the old one verbose showcase ipynb

add  demonstrating DrugEx with shape-based ROCS scoring.
 RDKit, CDPKit, and OpenEye ROCS implementations in multi-objective RL workflow.

Features:
- Fine-tuning pretrained RNN on CCR2 ligand data
- Multi-objective RL combining ROCS shape similarity and synthetic accessibility
- Molecule generation with all three ROCS backends
- Comprehensive results showing 92-96.5% success rates

fulopjoz committed:
refactor(scorers): deprecate max_centers in favor of max_isomers

- Add deprecation warn- Rename to max_isomers in
- RDKitConformerGeneratoring for max_centers parameter
- Rename to max_isomers in CDPKitConformerGenerator
- Update all docstrings and attribute documentation

fulopjoz committed:
perf(scorers): enable multi-conformer shape generation in CDPKit ROCS

  - Enable multiConformerMode(True) to generate shapes for all conformers
  - Change _generate_shape_helper to return list of shapes instead of single
   shape
  - Update worker to evaluate all query shapes against references
  - Improve best score tracking logic in nested loops

fulopjoz committed:
feat(scorers): add multi-reference group support and optimize RDKit ROCS
  scorer

  - Add support for Dict[str, List[str|Chem.Mol]] reference formats with
  named groups
  - Implement group-based scoring architecture with group_to_indices mapping
  - Extract _score_single_reference() helper function for better code reuse
  - Add _ensure_reference_conformers() to auto-generate missing conformers
  - Refactor worker processes to use global settings initializer pattern
  - Add _deduplicate_smiles() to avoid redundant conformer generation
  - Change _load_molecule_from_file to _load_molecules_from_file (loads all  mols)
  - Update getKey() to return group-specific or aggregate score names
  - Improve reference validation with detailed error messages
  - Return multi-column score arrays (one column per reference group)

fulopjoz committed:
fix(scorers): initialize scores_unique array to prevent parallel processing fallback

Fixed UnboundLocalError in RDKitROCSScorer.getScores() where scores_unique was referenced before assignment in the parallel processing path. This caused all parallel work to be discarded and recomputed sequentially

Added initialization of scores_unique array before the if/else branch to ensure it exists for both sequential and parallel code paths.

martin-sicho commented:
Hi @fulopjoz, I looked at the tutorial and I still think it is quite complex. I would suggest to do the following:

1. move the `rocs` folder under the `advanced` tutorial section.
2. Split the notebook into multiple ones
   - In particular, I would aim to have one quick start notebook that just shows the RL loop and only the necessary steps to do it showcasing only one open-source shape matching method 
     - I would move the fine-tuning code to a separate file that people can just run to get the input models for the RL notebook. It can also even just be a script. We can just point the reader to the already existing basic notebooks to get a grasp on what is done prior to RL.
   - You can move the other shape-matching methods to similar files dedicated to each method and discuss the specifics there
4. Add a readme file in the folder suggesting the order and structure of the tutorial notebooks that people should consider them in and point them to basic tutorials first as well.

I think this way it will be easier to follow than navigating one large notebook and you can easily focus only on the most important parts for shape matching.

fulopjoz committed:
--h

HellevdM committed:
small fixes in SchrodingerConformerGenerator file handeling

HellevdM committed:
fix stereoenumeration SchrodingerConformerGenerator

fulopjoz committed:
Merge pull request #1 from HellevdM/feature/rocs-scoring

small fixes in SchrodingerConformerGenerator file handeling

fulopjoz committed:
refactor(scorers): add multi-reference grouping and memory-efficient scoring to CDPKit ROCS

- Add dict-based reference grouping for multi-target optimization
- Implement SMILES deduplication for efficient batch processing
- Replace conformer passing with on-demand reloading in workers
- Enable multi-conformer mode for comprehensive shape evaluation
- Support group-based scoring across multiple reference sets

fulopjoz committed:
refactor(scorers): add multi-reference grouping and improve RDKit ROCS scoring

- Add dict-based reference grouping support for multi-target scenarios
- Implement SMILES deduplication to avoid redundant conformer generation
- Refactor worker initialization to use shared state pattern
- Add auto-conformer generation for 2D reference molecules
- Improve error handling and scoring logic for group-based evaluation

fulopjoz committed:
refactor(scorers): rename max_centers to max_isomers in conformer generators

- Rename max_centers parameter to max_isomers for clarity
- Add deprecation warning for backward compatibility
- Update documentation to reflect stereoisomer terminology
- Apply consistent naming across RDKit and CDPKit implementations

fulopjoz committed:
feat(data): add  active and decoy datasets for CCR2 threshold optimization

-  actives_ccr2_N75.csv containing 75 active compounds
-  decoys_ccr2_N500.csv  500 decoy compounds

fulopjoz committed:
refactor(config): update conformer generation parameters and paths

- Increased MAX_CONFORMERS from 20 to 50, MAX_ISOMERS from 2 to 4, and MAX_HEAVY_ATOMS from 30 to 45
- Added MAX_ROTATABLE_BONDS parameter to match OpenEye filter for flexibility.
- Updated ROCS_THRESHOLD and SA_THRESHOLD values for improved scoring accuracy.
- Refactored model path handling to use Path for better compatibility.

fulopjoz committed:
feat(tutorial): enhance ROCS quickstart notebook with threshold validation

- Updated tutorial to include ROCS threshold validation for shape similarity scoring.
- Improved clarity in tutorial descriptions and prerequisites.
- Included visualizations for threshold validation

fulopjoz committed:
feat(tutorial): add ROCS threshold analysis script for CCR2 ligand evaluation

- script for comprehensive ROCS threshold determination analysis.
- Implemented functions for loading datasets, scoring molecules, performing ROC analysis, and generating visualizations.
- Added capabilities for comparing thresholds and generating detailed reports on scoring performance.

fulopjoz committed:
fix(config): adjust RL training parameters

fulopjoz committed:
docs(tutorial): streamline ROCS tutorial for CCR2 ligand generation

- Simplified the tutorial structure to focus on a 3-step workflow: fine-tuning, RL training, and molecule generation.
- Enhanced quick start instructions and output descriptions for better user experience.

fulopjoz committed:
refactor(tutorial): extract functions and add type hints in generate_molecules

  Extract load_model(), generate_and_score(), and print_summary() functions.
  Add type hints and docstrings. Update notebook reference to rocs_rl_tutorial.

fulopjoz committed:
refactor(tutorial): modularize prepare_models with function extraction

  Extract load_and_standardize_data(), encode_smiles(), create_train_test_split(),
  and fine_tune_model() functions. Add type hints and step-based progress output.

fulopjoz committed:
docs(tutorial): add workflow documentation and threshold analysis guide

  Add notebook workflow section, threshold analysis explanation, configuration
  reference table, and enhanced troubleshooting. Improve navigation structure.

fulopjoz committed:
refactor(tutorial): rename quickstart_rl_rdkit to rocs_rl_tutorial
refactor(tutorial): remove obsolete ROCS-RNN-Demo notebook

  -better reflects comprehensive RL workflow tutorial content covering threshold
  validation, training, and analysis
  -functionality consolidated into rocs_rl_tutorial.ipynb

fulopjoz committed:
refactor(tutorial): extract functions and add type hints in CDPKit script

Extract create_cdpkit_environment(), run_rl_training(), and generate_molecules() Add type hints, rename CDPKIT_READY to CDPKIT_AVAILABLE.

fulopjoz committed:
refactor(tutorial): extract functions and add type hints in OpenEye script

  Extract create_openeye_environment(), run_rl_training(), and generate_molecules().
  Add type hints, rename OPENEYE_READY to OPENEYE_AVAILABLE.

fulopjoz committed:
fix(tutorial): ensure current threshold is included in analysis

fulopjoz committed:
Merge feature/rocs-update into feature/rocs-scoring

Resolved conflicts by accepting feature/rocs-update versions:
- conformer_generators.py: cleaner import order
- cdpkit_rocs.py: enhanced documentation
- rdkit_rocs.py: enhanced documentation
- ROCS-RNN-Demo.ipynb: updated notebook version

All tutorial refactoring from feature/rocs-update merged successfully.

fulopjoz committed:
docs(tutorial): update CDPKit  backend installation instructions

Clarified installation command

fulopjoz commented:
Hi @martin-sicho,  I made some changes: 

1. Advanced section: tutorial is at tutorial/advanced/rocs/
2. Simplified structure:
  - rocs_rl_tutorial.ipynb - main RL workflow (threshold validation + training)
  - prepare_models.py - separate fine-tuning script (users just run it)
  - generate_molecules.py - separate generation script
  - threshold_analysis.py -  separate script to find the optimal threshold

3. Method-specific files:
  - run_cdpkit_rocs.py - CDPKit implementation
  - run_openeye_rocs.py - OpenEye implementation
  - Notebook focuses on RDKit (open-source quickstart)
  
  4. README: Complete with tutorial order, workflow steps, and links to basic tutorials

fulopjoz committed:
chore(tutorial): remove generated outputs and comparison data

fulopjoz committed:
chore(tutorial): remove generated outputs and comparison data v2

fulopjoz committed:
fix(conformers): add num_threads parameter to control RDKit CPU usage

  - Added num_threads parameter to RDKitConformerGenerator.__init__
  - Updated _create_fresh_etkdg() to use configurable threading
  - Set num_threads=1 in config.py and threshold_analysis.py

fulopjoz committed:
style(scorers): move RDKit import to module level

fulopjoz committed:
docs(tutorial): add ROCS initialization examples

fulopjoz committed:
fix(conformers): initialize CDPKit molecules with calcBasicProperties

fulopjoz committed:
docs(scorers): clarify CLI ROCS CLI usage

fulopjoz committed:
refactor(conformers): remove tempdir copies in write_conformers

fulopjoz committed:
refactor(cdpkit): replace global variables with class-based worker pattern

 - Add CDPKitWorkerContext dataclass for typed worker state
 - Add CDPKitScoringWorker class with initialize() and __call__()
 - Remove _CDPKIT_WORKER_SETTINGS global variable
 - Remove _cdpkit_worker_init() and _score_molecule_cdpkit_worker()
 - Update getScores() to use new worker pattern

fulopjoz committed:
chore(.gitignore): clean up ROCS tutorial generated outputs section

fulopjoz committed:
chore(.gitignore): remove commented-out ROCS tutorial generated outputs

fulopjoz committed:
docs(tutorial): show conformer generator and scorer initialization for all backends

- Add CDPKit, OpenEye backend example alongside RDKit in notebook
- Document reference ligand formats (.sdf, .mol2, .sq) in README
- Add Shape Query (.sq) file explanation and examples
-Link notebook to README

fulopjoz committed:
refactor(conformers): remove deprecated code and comments
- Remove max_centers parameter and deprecation warning, comments,  redundant imports, warnings

fulopjoz committed:
feat(scorers): add OpenEyeROCSScorer implementation

-- Rename CLIROCSScorer to OpenEyeROCSScorer
- Supports .sdf and .sq (Shape Query) reference formats

fulopjoz committed:
fix(tutorial): update imports after scorer renaming

-CDPKit: cdpkit_rocs -> rocs_cdpkit
-OpenEye: cli_rocs.CLIROCSScorer -> rocs_openeye.OpenEyeROCSScorer

fulopjoz committed:
perf(tutorial): optimize conformer generation
  threading

config.py:
 - num_threads=0 (all cores)

 threshold_analysis.py:
  - Add: num_threads=1 when n_jobs=1

fulopjoz committed:
docs(tutorial): update README and notebook for reference ligand formats, all backends - setup

fulopjoz committed:
drugex/training/scorers/rdkit_rocs.py (renamed) to rocs_rdkit.py

fulopjoz committed:
(refactor): simplify threshold workflow

fulopjoz committed:
(refactor): rename OpenEye ROCS query_files to references

fulopjoz committed:
rename file

fulopjoz committed:
update impors

HellevdM commented on drugex/training/scorers/rocs_openeye.py:
```diff
@@ -0,0 +1,364 @@
+import os
```
```suggestion
"""OpenEye-ROCS based ROCS scorer."""
import os
```

For consistency with the other ROCS scorers

HellevdM commented on drugex/training/scorers/rocs_openeye.py:
```diff
@@ -0,0 +1,364 @@
+            output_file,
+            "-prefix",
+            "rocs",
+            "-outputdir",
+            output_dir,  # Add output directorys
+            "-rankby",
+            self.score_type,
+            "-chemff",
+            self.color_force_field,
+            "-cutoff",
+            "-1.0",  # Return all molecules (no cutoff)
+            "-maxhits",
+            "0",  # Return all molecules (overrides besthits)
+            "-tanimoto_cutoff",
+            "0.0",
+            "-stats",
+            "best",
+            "-nostructs",
+            "-scdbase",  # Don't combine contiguous conformers
+        ]
+
+        # Add shapeonly explicitly (with true/false value) instead of conditionally
+        cmd.extend(["-shapeonly", str(self.shape_only).lower()])
+
+        # Add opt explicitly instead of conditionally
+        cmd.extend(["-opt", str(self.optimize).lower()])
+
+        # Add optchem explicitly with proper condition
+        if not self.shape_only and self.color_optimize:
+            cmd.extend(["-optchem", "true"])
+        else:
+            cmd.extend(["-optchem", "false"])
+
+        return cmd
```
```suggestion
    def _build_rocs_command(
        self, query_file: str, input_file: str, output_file: str
    ) -> List[str]:
        """Build ROCS command matching cli_base_rocs.py exactly"""
        # Extract the directory from output_file
        output_dir = os.path.dirname(output_file) or "."

        cmd = [
            self.binary_path,
            "-query", query_file,
            "-dbase", input_file,
            "-report", "one",
            "-reportfile", output_file,
            "-prefix", "rocs",
            "-outputdir", output_dir,
            "-stats", "best", # Include best overlay(s) for each dbase molecule
            "-nostructs", # Don't save the structures
            "-scdbase",  # Don't combine contiguous conformers
        ]
        
        if self.shape_only:
            # sets chemff none, optchem false and rankby tanimoto
            cmd.extend(["-shapeonly", str(self.shape_only).lower()])
        else:
            cmd.extend(["-rankby", self.score_type])
            cmd.extend(["-chemff", self.color_force_field])
            
        # set optimizer on/off, if off score only
        cmd.extend(["-opt", str(self.optimize).lower()])

        return cmd
```

Improve readability of rocs command by moving key value pairs to the same line and removing arguments (`cutoff`, `maxhits`, `tanimoto_cutoff`) which had default values. As well as improve readability of the flow of setting `shape_only`, `color_optimize` and `opt`.

HellevdM commented on drugex/training/scorers/rocs_openeye.py:
```diff
@@ -0,0 +1,364 @@
+    - Hybrid Python/CLI workflow for flexible integration
+
+    Attributes:
+        - references: dict of query files for ROCS queries (.sq or molecule files).
```
```suggestion
    Attributes:
        - conformer_generator: conformer generator used for generated molecules.
        - references: dict of query files for ROCS queries (.sq or molecule files).
```

HellevdM commented on drugex/training/scorers/rocs_openeye.py:
```diff
@@ -0,0 +1,364 @@
+            is returned for that key.
+        - score_type: Type of scoring to use (e.g., TanimotoCombo)
+        - shape_only: If True, only shape scoring is performed
+        - optimize: If True,
+        - color_optimize: If True, color optimization is performed
+        - color_force_field: Force field to use for color optimization
+        - rocs_binary: Name of the ROCS binary to use
+        - binary_path: Path to the ROCS binary (if not in PATH)
+        - show_progress: If True, progress is shown during scoring
+        - name_suffix: Optional suffix for the scorer name (important if multiple
+           OpenEyeROCSScorer scorers are used in the same environment)
+    """
+
+    def __init__(
+        self,
+        conformer_generator: ConformerGenerator,
+        references: dict[str, List[str] | str],
+        score_type: str = "TanimotoCombo",
+        shape_only: bool = False,
+        optimize: bool = True,
+        color_optimize: bool = True,
+        color_force_field: str = "ImplicitMillsDean",
+        rocs_binary: str = "rocs",
+        binary_path: str | None = None,
+        show_progress: bool = True,
+    ):
+
+        super().__init__()
```
```suggestion
    ):
        """Initialize the OpenEye ROCS scorer.
        
        Args:
            conformer_generator: generator used to produce generated molecule conformers
            references: dict mapping scorer name to SDF/sq query name(s). If multiple 
                dict items are specified, the output will be a score per item. It a dict
                item has multiple paths specified, the molecules will be scored for each 
                query but only the maximum of the scores will be returned (i.e. the best
                match).  
            score_type: Type of scoring to use (e.g., TanimotoCombo), ignored
                if shape_only
            shape_only: If True, only shape scoring is performed
            optimize: turn optimizer on/off, if off score only
            color_optimize: If True, color optimization is performed
            color_force_field: : Force field to use for color optimization
            rocs_binary: Name of the ROCS binary to use
            binary_path: Path to the ROCS binary (if not in PATH)
            show_progress: If True, progress is shown during scoring
        
        Raises:
            ImportError: If OpenEye toolkits are not available.
            FileNotFoundError: If rocs binary not found
        """
        super().__init__()
```

HellevdM reviewed (commented):
Hi @fulopjoz,

Thanks a lot for implementing all these changes!
I had a look at the updated files and have a few more minor suggestions to add:
- I think it would be nice to add a bit more explanation in the tutorial file, e.g. some example of the reference compounds you used and perhaps a few generated compounds. Related to this, I think the generation part could still be in the jupyter notebook tutorial file.
- Maybe the names of the arguments and attributes for the references in the ROCS scorers could be standardized. Creating a multi-target scorer is now referred to as "groups" in the `CPDKitROCSScorer` and `RDKitROCSScorer`, but as "queries" in the `OpenEyeRocsScorer`. Additionally, "query molecules" is used to refer to the molecules to be scored in the in the `CPDKitROCSScorer` and `RDKitROCSScorer`, but as mentioned "query" is used to refer to the reference compounds in the `OpenEyeROCSScorer`.
- I made a few code suggestions for mainly the documentation of the `OpenEyeROCSScorer`

fulopjoz committed:
Update drugex/training/scorers/rocs_openeye.py

Co-authored-by: H.W. van den Maagdenberg <h.w.van.den.maagdenberg@lacdr.leidenuniv.nl>

fulopjoz committed:
Update drugex/training/scorers/rocs_openeye.py

Co-authored-by: H.W. van den Maagdenberg <h.w.van.den.maagdenberg@lacdr.leidenuniv.nl>

fulopjoz committed:
Update drugex/training/scorers/rocs_openeye.py

Co-authored-by: H.W. van den Maagdenberg <h.w.van.den.maagdenberg@lacdr.leidenuniv.nl>

fulopjoz committed:
Update drugex/training/scorers/rocs_openeye.py

Co-authored-by: H.W. van den Maagdenberg <h.w.van.den.maagdenberg@lacdr.leidenuniv.nl>

fulopjoz committed:
docs(rocs): standardize naming to use "reference groups" instead of "queries"

fulopjoz committed:
feat(tutorial): add molecule generation section to ROCS RL notebook

fulopjoz committed:
docs(tutorial): note that molecule generation is demonstrated in notebook

fulopjoz commented:
Hi @HellevdM, 

Thanks for the feedback! I have addressed your suggestions:
- standardized OpenEyeROCSScorer to use "reference groups" terminology (matching CDPKit/RDKit), only docs updated
- added molecule generation section to the tutorial notebook with example output
- updated README to note generation is now in the notebook

martin-sicho reviewed (approved):
Just checked everything and it looks good. I will just re-run the tutorial now to see if it works and will merge if everything is OK there.
